backend/msgpack/server.go

```diff
@@ -182,12 +194,45 @@ func (s *server) write(
 		s.RUnlock()
 		return err
 	}
-	numShards := placement.NumShards()
-	shard := s.shardFn(id, numShards)
-	instances := placement.InstancesForShard(shard)
-	err = s.writerMgr.WriteTo(instances, shard, mu, pl)
+	var (
+		numShards = placement.NumShards()
+		shardID   = s.shardFn(id, numShards)
+		instances = placement.InstancesForShard(shardID)
+		nowNanos  = s.nowFn().UnixNano()
+		multiErr  = xerrors.NewMultiError()
+	)
+	for _, instance := range instances {
+		shard, ok := instance.Shards().Shard(shardID)
+		if !ok {
+			err := fmt.Errorf("instance %s does not own shard %d", instance.ID(), shardID)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+		writeEarliestNanos, writeLatestNanos := s.writeTimeRangeFor(shard)
+		if nowNanos < writeEarliestNanos || nowNanos > writeLatestNanos {
```

Comment:
maybe it's better to have a function called `shouldWriteToShard(shard, nowNanos) bool` which takes care of the writeTimeRangeFor() + nowNanos check?